[Richard Dominick] iP

RichDom2185 opened the pull request:
# Mia

Mia is an assistant that has some of the following quality-of-life updates to the CLI:
* All messages are wrapped in chat bubbles 💬 
* Automatically breaks long inputs across multiple lines
* _...and many more..._

How to run:
1. Make sure you have Java 11 installed
1. Head over to the [`master`](/RichDom2185/ip/tree/master) branch and find download the latest `.jar` file from the "Releases" tab
1. Run the `.jar` file

~~Make sure you run it from a normal terminal window, not IntelliJ's integrated terminal.~~ In order to support ANSI escape codes (to overwrite the text prompt with a chat bubble), please **use a terminal emulator that supports it**.

Todo:
- [x] Replace `String` with `Span`
- [ ] Support more rich text types over the CLI
- [ ] Add GUI
- [ ] Deprecate CLI 😢 

## The `Text` Interface

> View the source code [here](/RichDom2185/ip/tree/master/src/main/java/general/ui/Text.java).

Originally meant to replace `String` in order to support overriding of `String::length` when using ANSI escape codes, so far only `Span` (direct substitute for a normal `String`) uses it due to lack of time to implement other classes.

```java
// ...
public interface Text {
    String getText();

    int length();

    Stream<? extends Text> lines();

    Text[] split(String regex);

    Text strip();
}
```

Commit:
Add Gradle support

RichDom2185 committed:
Rename chatbot to Mia

RichDom2185 committed:
Add response printer

RichDom2185 committed:
Fix response printer bugs

RichDom2185 committed:
Add echo and exit

RichDom2185 committed:
Improve code quality and UI

RichDom2185 committed:
Improve text handling abstraction

RichDom2185 committed:
Add left padding to right-align response bubbles

RichDom2185 committed:
Use chat bubbles for entered commands

RichDom2185 committed:
Create Task, TaskManager classes

RichDom2185 committed:
Add task list features

* Update string representation of tasks and task manager
* Update default behaviour of chatbot on input
* Miscellaneous changes

RichDom2185 committed:
Use a wrapper class for text items

RichDom2185 committed:
Allow toggling of task status

RichDom2185 committed:
Add multiple task types

RichDom2185 committed:
Add ability to create multiple task types

Also restore echo function for unknown commands

RichDom2185 committed:
Update test scripts

RichDom2185 committed:
Update expected test case output

RichDom2185 committed:
Add some error handling to incorrect number of arguments

RichDom2185 committed:
Add ability to delete tasks

RichDom2185 committed:
Merge remote-tracking branch 'origin/week2'

RichDom2185 committed:
Improve code quality

Refactor some code and make Mia multiple-instantiable.

RichDom2185 committed:
Refactor chat bubbles into a separate ChatWindow object

RichDom2185 committed:
Encapsulate prompt under ChatWindow object

Also fix bug regarding missing prompt on syntax error for deadlines and events.

RichDom2185 committed:
Improve code quality

RichDom2185 committed:
Update Javadoc formatting

RichDom2185 committed:
Improve abstraction of handling responses

RichDom2185 committed:
Add Command class

RichDom2185 committed:
Save data to disk

RichDom2185 committed:
Load save data from disk

RichDom2185 committed:
Update gitignore

RichDom2185 committed:
Add date and time parsers

RichDom2185 committed:
Update date handling in Deadline to use date objects

RichDom2185 committed:
Update date handling in Event to use date objects

RichDom2185 committed:
Merge remote-tracking branch 'origin/branch-Level-7' into week3

RichDom2185 committed:
Merge remote-tracking branch 'origin/branch-Level-8' into week3

RichDom2185 committed:
Fix compile errors + Update save file format post-merge

RichDom2185 committed:
Update invalid state message

RichDom2185 committed:
Merge branch 'week3' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into refactor-command-handling

RichDom2185 committed:
Add context "callbacks" to Command class

RichDom2185 committed:
Migrate exit handling to use Command abstraction

RichDom2185 committed:
Fix bug in argument parsing

RichDom2185 committed:
Migrate command handling logic to use Command abstraction

RichDom2185 committed:
Update error messages

RichDom2185 committed:
Merge branch 'refactor-command-handling' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week3

RichDom2185 committed:
Support multiple different save files in multiple instances

RichDom2185 committed:
Move RegexHelper into general utils package

RichDom2185 committed:
Move UI-related classes into general UI package

RichDom2185 committed:
Move task-related classes to Mia's package for supported tasks

RichDom2185 committed:
Move Mia-related code to Mia package

RichDom2185 committed:
Revert "Move task-related classes to Mia's package for supported tasks"

This reverts commit 30b0c3f000d29721fc1e6c7c3463be84bf4dc9da.

RichDom2185 committed:
Move task-related classes to Mia package

RichDom2185 committed:
Update test script file permissions

RichDom2185 committed:
Fix test scripts post-refactoring

RichDom2185 committed:
Add TODOs

RichDom2185 committed:
Create initial StoredFile class

RichDom2185 committed:
Migrate file reading to StoredFile methods

RichDom2185 committed:
Migrate file writing to StoredFile methods

RichDom2185 committed:
Merge branch 'refactor-storage-handling' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week3

RichDom2185 committed:
Merge branch 'use-packages' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week3

RichDom2185 committed:
Move StoredFile into general utils package

RichDom2185 committed:
Merge remote-tracking branch 'source/add-gradle-support' into week3

RichDom2185 committed:
Update Gradle build configuration

RichDom2185 committed:
Fix broken gradle build config bug

RichDom2185 committed:
Remove unused setter in Task

RichDom2185 committed:
Add tests for ChatWindow

RichDom2185 committed:
Add tests for Todo

RichDom2185 committed:
Add tests for Event

RichDom2185 committed:
Add tests for Deadline

RichDom2185 committed:
Add documentation to some classes

RichDom2185 committed:
Add author tag to Span class

RichDom2185 committed:
Add documentation to Task, TaskManager classes

RichDom2185 committed:
Add documentation to StoredFile class

RichDom2185 committed:
Replace wildcard with explicit imports

RichDom2185 committed:
Refactor task list string representation generator to own method

RichDom2185 committed:
Add methods to check for search query match

RichDom2185 committed:
Add new Command to search for tasks

RichDom2185 committed:
Fix empty search bug

RichDom2185 committed:
Merge branch 'branch-Level-9' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week3

RichDom2185 committed:
Add line breaks

RichDom2185 committed:
Update binary operator position

RichDom2185 committed:
Update binary operator position in Mia logo

RichDom2185 committed:
Rearrange methods

RichDom2185 committed:
Merge branch 'branch-A-CodingStandard' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week3

RichDom2185 committed:
Add Task-related documentation for parsing a save format

RichDom2185 committed:
Add documentation for Mia class

RichDom2185 committed:
Add documentation for Command class

RichDom2185 committed:
Add documentation for StoredFile class

RichDom2185 committed:
Update documentation for RegexHelper class

RichDom2185 committed:
Add documentation for Task class

RichDom2185 committed:
Update documentation for TaskManager class

RichDom2185 committed:
Merge branch 'branch-A-JavaDoc' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week3

RichDom2185 committed:
Add documentation for newly implemented methods post-merge

TZL0 commented on src/main/java/general/ui/Text.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.stream.Stream;
+
+public interface Text {
```
I like your class Text to process String in the UI.

RichDom2185 replied:
Thanks! I created the interface mainly because I had envisioned for proper rich text support, so whatever that is passed in may not be a pure `String`, so I wanted to make a common standard `Text` that was understandable by the app as something to parse and print accordingly

TZL0 commented on src/main/java/general/ui/ChatWindow.java:
```diff
@@ -0,0 +1,80 @@
+        sc = new Scanner(System.in);
+    }
+
+    private static Text[] breakLines(Text data) {
```
Is it better to name it getBrokenLines()?

jwdavis0200 replied:
I agree, according to the CS2103T coding standards such a getter method should have the verb infront of it :-)
Interesting work though with this method!

RichDom2185 replied:
Hmm, could you point me to which part specifically? This isn't a getter method by any means, as it takes in a `Text` and returns a broken-up version of it. It does not depend on any internal object state at all.

TZL0 commented on src/main/java/mia/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+        this(title, false, deadline);
+    }
+
+    Deadline(String title, boolean isCompleted, String deadline) throws IllegalArgumentException {
```
Is it better to add an access modifier?

RichDom2185 replied:
Good catch, will add!

TZL0 commented on src/main/java/mia/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+    private LocalDate byDate;
+    private LocalTime byTime;
+
+    Deadline(String title, String deadline) throws IllegalArgumentException {
```
Is it better to add an access modifier?

RichDom2185 replied:
Good catch, will add!

TZL0 commented on src/main/java/mia/Event.java:
```diff
@@ -0,0 +1,69 @@
+    private LocalDate endDate;
+    private LocalTime endTime;
+
+    Event(String title, boolean isCompleted, String time) {
```
Is it better to add an access modifier?

RichDom2185 replied:
Good catch, will add!

TZL0 commented on src/main/java/mia/Event.java:
```diff
@@ -0,0 +1,69 @@
+        this(title, isCompleted, time, time);
+    }
+
+    Event(String title, String time) {
```
Is it better to add an access modifier?

TZL0 commented on src/main/java/mia/Mia.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @return The {@code TaskManager} associated with this {@code Mia} instance.
+     */
+    public TaskManager getTasks() {
```
Why not name it getTaskManager() to better identify return type?

RichDom2185 replied:
Because I felt that the return type should not influence the method name. In my opinion, the method name should return what the object _is_ (e.g. "get my laptop"), in natural language (to be as clear as possible), as opposed to specifying the exact specification (e.g. "get my HP Book Series-12403", as this is provided from the method signature anyway.

What do you think?

_PS: The HP model is fake (I think)_

jwdavis0200 commented on src/main/java/general/ui/ChatWindow.java:
```diff
@@ -0,0 +1,80 @@
+                .toArray(Text[]::new);
+    }
+
+    public void dispose() {
```
Hmm, since it's just a single line method call, would it be better just to skip the implementation of this as a method since you would still have to call the method anyways. Could it affect the number of stack calls the program have to call? Although performance difference would be really negligible but something to think about.

RichDom2185 replied:
This (using a wrapper method) is the only way to call it without exposing the entire internals of how `ChatWindow` works. In terms of number of stack calls, I would argue that leaving the resource (in this case, `Scanner`) open when we're no longer using it is worse than an extra function call. Plus, compiler does its own optimisations. What do you think?

jwdavis0200 commented on src/main/java/general/ui/ChatWindow.java:
```diff
@@ -0,0 +1,80 @@
+                maxLength = lines[i].length();
+            }
+        }
+        final int paddingLength = windowWidth - maxLength - 4;
```
Great use of constant here, since padding length won't be changing.

RichDom2185 replied:
Thanks!

jwdavis0200 commented on src/main/java/general/utils/StoredFile.java:
```diff
@@ -0,0 +1,123 @@
+    public static StoredFile from(String filePath) {
+        final File file = new File(filePath);
+        if (file.exists()) {
+            return new ExistentFile(filePath);
```
Good naming scheme, since file exists, the class is an ExistentFile!

RichDom2185 replied:
Thanks!

jwdavis0200 commented on src/main/java/mia/Command.java:
```diff
@@ -0,0 +1,254 @@
+     * @return A {@code Command} for the context to call
+     * @throws IllegalArgumentException When a parsing error occurs
+     */
+    public static Command from(Mia context, String input) throws IllegalArgumentException {
```
No problem here, but seeing you're really proficient, why not simplify your code by implementing enums? :-)

RichDom2185 replied:
I don't see the need for having a separate enum for every single command type. I like to think of it as: what if in the future you want to have another command that exits while not saving any changes (for example)? What about other behaviour? It would not make sense to use multiple conditional checks for every single enum type.

But by using "flags" such as `shouldCloseContext`, or potentially `shouldXXXX` in the future, this makes it easier. Also using enums breaks abstraction. Why should a the `context` the command is being run in know about what types of `Command`s are available?

jwdavis0200 commented on src/main/java/mia/TaskManager.java:
```diff
@@ -0,0 +1,168 @@
+                            }
+                            return Stream.empty();
+                        }
+                ).collect(Collectors.toList());
```
Impressive work with the Java Collections Framework!

RichDom2185 replied:
Thanks!

jwdavis0200 commented on src/main/java/mia/TaskManager.java:
```diff
@@ -0,0 +1,168 @@
+     * @param query The query to be word-matched against the task titles
+     * @return A {@code String} representation of the tasks that matched the given query
+     */
+    public String search(String query) {
```
Since your method will be searching AND returning the tasks, maybe searchResult would be a better name?

RichDom2185 replied:
Hmm I'm not too sure what you are trying to convey, as I don't really find `searchResult` that natural-sounding. The method name should start with a verb right, so `searchSomething`. Maybe you mean `searchTasks`?

jwdavis0200 commented on src/test/java/TodoTest.java:
```diff
@@ -0,0 +1,20 @@
+        final Todo todo = Todo.fromSaveFormat(DUMMY_TODO);
+        assertEquals("📜 ☐ Title", todo.toString());
+    }
+}
```
I love your extensive testing of your program, this is definitely a very good habit for future development projects!

RichDom2185 replied:
Thanks!

jwdavis0200 reviewed (commented):
Really impressive work!! I have left some comments on some potential coding standard issues, really minor ones. Your program has been one that's very interesting to review. There are also comments left by another reviewer, whose comments I agree with. Otherwise LGTM, great job!

RichDom2185 committed:
Merge branch 'week3' of https://github.com/RichDom2185/nus-cs2103t-2210-ip

RichDom2185 committed:
Add missing access modifiers

RichDom2185 committed:
Update constructor body

RichDom2185 committed:
Add checkstyle

RichDom2185 committed:
Update gitignore

RichDom2185 committed:
Fix checkstyle issues

RichDom2185 committed:
Set up boilerplate for initial GUI

RichDom2185 committed:
Add avatars

RichDom2185 committed:
Add pre-commit hooks

RichDom2185 committed:
Configure Makefile for hooks and lints

RichDom2185 committed:
Fix lint issues

RichDom2185 committed:
Fix lint configurations

RichDom2185 committed:
Merge branch 'add-hooks' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week4

RichDom2185 committed:
Merge branch 'add-gui' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week4

RichDom2185 committed:
Merge branch 'week4' of https://github.com/RichDom2185/nus-cs2103t-2210-ip

RichDom2185 committed:
Command class: empty A-Assertions commit

Assertions are already being used in the Command class.

This is already the case since commit
796ab22c907149a120fe0833bc0db6bcb5f49977.

Committing an empty commit allows for the history to diverge from the
main branch. That way, a PR can be created for this branch. The PR is
needed as it is one of the requirements for this module.

In addition, this full commit description is also necessary to satisfy
another requirement for full commit messages in this module.

Let's create an empty commit to fulfil these requirements.

RichDom2185 committed:
Fix grammar in header comment

RichDom2185 committed:
Improve abstraction level of Mia command execution

RichDom2185 committed:
Span class: empty A-Streams commit

Streams are already being used in the Span class.

This is already the case since commit
4de4b19d08a609069b5d8206cfaba4960de1a66b.

Committing this empty commit allows for the history to diverge from the
main branch. That way, a PR can be created for this branch. The PR is
needed as part of the requirements for this module.

In addition, this full commit description is also necessary to satisfy
another requirement for this module, which are having some commits with
full descriptions.

Let's create an empty commit to fulfil these requirements.

RichDom2185 committed:
TaskManager class: empty A-Streams commit

Streams are already being used in the TaskManager class.

This is already the case since commit
a94055ad755806a79832f5718c5addffe66feda7.

Committing an empty commit allows for the history to diverge from the
main branch. That way, a PR can be created for this branch, which
fulfils one of the requirements for this module.

In addition, this full commit descripton also satisfies another module
requirement which are having some commits with full commit descriptions.

Let's create an empty commit to satisfy these requirements.

RichDom2185 committed:
Merge pull request #2 from RichDom2185/branch-A-Assertions

Add some assertions to codebase

RichDom2185 committed:
Merge branch 'week5' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into branch-A-CodeQuality

RichDom2185 committed:
Merge branch 'week5' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into branch-A-Streams

RichDom2185 committed:
Merge pull request #4 from RichDom2185/branch-A-Streams

Use Streams in codebase where applicable

RichDom2185 committed:
Merge branch 'week5' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into branch-A-CodeQuality

RichDom2185 committed:
Add tool versions

RichDom2185 committed:
Add FXML files for Mia GUI

RichDom2185 committed:
Duplicate Mia output to ArrayList buffer

RichDom2185 committed:
Add controllers for FXML files

RichDom2185 committed:
Implement initial GUI for Mia

RichDom2185 committed:
Expose state flags publicly

Allows Mia to be controlled from external context managers (JavaFX).

RichDom2185 committed:
Merge branch 'branch-Level-10' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into week5

RichDom2185 committed:
Merge branch 'week5' of https://github.com/RichDom2185/nus-cs2103t-2210-ip into branch-A-CodeQuality

RichDom2185 committed:
Add punctuation to documentation

RichDom2185 committed:
Merge pull request #3 from RichDom2185/branch-A-CodeQuality

Improve quality of codebase

RichDom2185 committed:
C-NaturalDates: Fix single-digit hour regex parsing error

RichDom2185 committed:
Merge branch 'week5' of https://github.com/RichDom2185/nus-cs2103t-2210-ip

RichDom2185 committed:
Update README

RichDom2185 committed:
Add user guide

RichDom2185 committed:
Add title to GUI window

RichDom2185 committed:
Merge branch 'week6' of https://github.com/RichDom2185/nus-cs2103t-2210-ip
